File: src/332_reconstruct_itinerary/ReconstructItinerary.java
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

class Solution {
    public List<String> findItinerary(List<List<String>> tickets) {
        Map<String, List<String>> adjList = new HashMap<>();
        for (List<String> ticket : tickets) {
            adjList.putIfAbsent(ticket.get(0), new ArrayList<>());
        }
        // Sort tickets in reverse order and add to adjList
        tickets.sort((a, b) -> b.get(1).compareTo(a.get(1)));
        for (List<String> ticket : tickets) {
            adjList.get(ticket.get(0)).add(ticket.get(1));
        }

        List<String> res = new ArrayList<>();

        dfs("JFK", adjList, res);

        Collections.reverse(res);
        return res;
    }

    private void dfs(String src, Map<String, List<String>> adjList, List<String> res) {
        while (adjList.containsKey(src) && !adjList.get(src).isEmpty()) {
            String dst = adjList.get(src).remove(adjList.get(src).size() - 1);
            dfs(dst, adjList, res);
        }
        res.add(src);
    }
}

public class ReconstructItinerary {
    public static void main(String[] args) {
        testFindItinerary();
    }

    public static void testFindItinerary() {
        Solution solution = new Solution();

        // Test case 1: Example 1
        List<List<String>> tickets1 = List.of(
                List.of("MUC", "LHR"),
                List.of("JFK", "MUC"),
                List.of("SFO", "SJC"),
                List.of("LHR", "SFO"));
        List<String> expected1 = List.of("JFK", "MUC", "LHR", "SFO", "SJC");
        List<String> actual1 = solution.findItinerary(tickets1);
        System.out.println(actual1 + " | " + expected1);

        // Test case 2: Example 2
        List<List<String>> tickets2 = List.of(
                List.of("JFK", "SFO"),
                List.of("JFK", "ATL"),
                List.of("SFO", "ATL"),
                List.of("ATL", "JFK"),
                List.of("ATL", "SFO"));
        List<String> expected2 = List.of("JFK", "ATL", "JFK", "SFO", "ATL", "SFO");
        List<String> actual2 = solution.findItinerary(tickets2);
        System.out.println(actual2 + " | " + expected2);

        // Test case 3: Single ticket
        List<List<String>> tickets3 = List.of(List.of("JFK", "ABC"));
        List<String> expected3 = List.of("JFK", "ABC");
        List<String> actual3 = solution.findItinerary(tickets3);
        System.out.println(actual3 + " | " + expected3);

        // Test case 4: Two tickets
        List<List<String>> tickets4 = List.of(
                List.of("JFK", "A"),
                List.of("A", "JFK"));
        List<String> expected4 = List.of("JFK", "A", "JFK");
        List<String> actual4 = solution.findItinerary(tickets4);
        System.out.println(actual4 + " | " + expected4);

        // Test case 5: Multiple from JFK
        List<List<String>> tickets5 = List.of(
                List.of("JFK", "A"),
                List.of("JFK", "B"),
                List.of("A", "JFK"));
        List<String> expected5 = List.of("JFK", "A", "JFK", "B");
        List<String> actual5 = solution.findItinerary(tickets5);
        System.out.println(actual5 + " | " + expected5);
    }
}

File: src/332_reconstruct_itinerary/reconstruct_itinerary.py
from collections import defaultdict
from typing import List


class Solution:
    def findItinerary(self, tickets: List[List[str]]) -> List[str]:
        adjList = defaultdict(list)
        for src, dst in sorted(tickets)[::-1]:
            adjList[src].append(dst)

        res = []

        def dfs(src):
            while adjList[src]:
                dst = adjList[src].pop()
                dfs(dst)

            res.append(src)

        dfs("JFK")
        return res[::-1]


# Test helper
def test_find_itinerary():
    solution = Solution()

    # Test case 1: Example 1
    tickets1 = [["MUC", "LHR"], ["JFK", "MUC"], ["SFO", "SJC"], ["LHR", "SFO"]]
    expected1 = ["JFK", "MUC", "LHR", "SFO", "SJC"]
    actual1 = solution.findItinerary(tickets1)
    print(f"{actual1} | {expected1}")

    # Test case 2: Example 2
    tickets2 = [
        ["JFK", "SFO"],
        ["JFK", "ATL"],
        ["SFO", "ATL"],
        ["ATL", "JFK"],
        ["ATL", "SFO"],
    ]
    expected2 = ["JFK", "ATL", "JFK", "SFO", "ATL", "SFO"]
    actual2 = solution.findItinerary(tickets2)
    print(f"{actual2} | {expected2}")

    # Test case 3: Single ticket
    tickets3 = [["JFK", "ABC"]]
    expected3 = ["JFK", "ABC"]
    actual3 = solution.findItinerary(tickets3)
    print(f"{actual3} | {expected3}")

    # Test case 4: Two tickets
    tickets4 = [["JFK", "A"], ["A", "JFK"]]
    expected4 = ["JFK", "A", "JFK"]
    actual4 = solution.findItinerary(tickets4)
    print(f"{actual4} | {expected4}")

    # Test case 5: Multiple from JFK
    tickets5 = [["JFK", "A"], ["JFK", "B"], ["A", "JFK"]]
    expected5 = ["JFK", "A", "JFK", "B"]
    actual5 = solution.findItinerary(tickets5)
    print(f"{actual5} | {expected5}")


if __name__ == "__main__":
    test_find_itinerary()
